### PR TITLE
Add first-class span types proposal

### PR DESCRIPTION
**Pull request by 333fred**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Question. If we have your this, can we remove the special roles rules we have with collections to prefer spans over stays arrays?   Would that just all fall out trivially once we have these conversions through preexisting better ness rules?  If so, that would be lovely.

**Comment by 333fred** [comment]
> Question. If we have your this, can we remove the special roles rules we have with collections to prefer spans over stays arrays? Would that just all fall out trivially once we have these conversions through preexisting better ness rules? If so, that would be lovely.

Potentially. I haven't worked through _all_ the scenarios, but I do believe it would make `Span` and `ReadOnlySpan` encompassing types in more scenarios, allowing the desired semantics to just fall out.

**Comment by CyrusNajmabadi** [comment]
Lovely!

File: proposals/first-class-span-types.md [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+We also add _implicit span conversion_ to the list of acceptable implicit conversions on the first parameter of an extension method when determining applicability
+([12.8.9.3](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md#12893-extension-method-invocations)) (change in bold):
+
+> An extension method `Cᵢ.Mₑ` is ***eligible*** if:
+> 
+> - `Cᵢ` is a non-generic, non-nested class
+> - The name of `Mₑ` is *identifier*
+> - `Mₑ` is accessible and applicable when applied to the arguments as a static method as shown above
+> - An implicit identity, reference ~~or boxing~~**, boxing, or span** conversion exists from *expr* to the type of the first parameter of `Mₑ`.
```
Comment (yaakov-h) [inline proposals/first-class-span-types.md:50]: The bold syntax isn’t working here on the GitHub renderer.

File: proposals/first-class-span-types.md [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+### Implicit Span Conversions
+
+We add a new type of implicit conversion to the list in [§10.2.1](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/conversions.md#1021-general), an
+_implicit span conversion_. This conversion is defined as follows:
+
+------
+
+An implicit span conversion permits `array_types`, `System.Span<T>`, `System.ReadOnlySpan<T>`, and `string` to be converted between each other as follows:
+* From any `array_type` with element type `Ei` to `System.Span<Ei>`
```
Comment (jnm2) [inline proposals/first-class-span-types.md:30]: Only single-dimensional `array_type`s have conversions to spans today.

**Pull request merged**